### ISSUE TITLE
vsphere: provision a vm from a template already installed?

### ISSUE THREAD

**Issue by sjpotter** [apcera/libretto#112]
I can easily create a template within vsphere, but it doesn't seem to be anyway to easily use libretto to provision it. it be nice if I could skip the ovf local file and just specify the template on the server.

**Comment by sjpotter** [comment]
i.e. I'd expect to be able to do something like this written in python

https://gist.github.com/snobear/8788977

**Comment by avnish30jn** [comment]
This is what libretto does now to provision a vm.

- Generates a template name, (vm.Template + "-" + selected_datastore_name) 
- checks if the vm with the template name exists or not
- if No, creates the VM with template name and mark it as template
- if yes, skips the template vm creation
- Clones the template VM to create the vm specified in the vm.Name and powers it on

Currently it doesn't add the ram, no of cpus, other vm configs to the clone vm configuration.

To make things work in libretto the below changes will be required. 

- Addition of two more configurations to cisp = types.VirtualMachineCloneSpec which is being used while cloning the vm
  1) VirtualMachineConfigSpec which will have ram, cpu and other information
  2) CustomizationSpec which will have the IP details

`	cisp := types.VirtualMachineCloneSpec{
		Location: relocateSpec,
		Template: false,
		PowerOn:  false,
	}`
- Generating the template name imposes a restriction on the name of the template. 
If it is not a use case, can we use vm.Template instead of generating the template name ?

**Comment by sjpotter** [comment]
So perhaps the correct way to go about this is

1) make a PR that enables using a precreated template instead of using an OVF file to upload and create it.  This wouldn't remove the current functionality, but would only work for one or the other (i.e. if both fields are specified, it would return an error)

2) another PR that provides more flexibility in terms of the configspec / customationspec.

thoughts?

**Comment by sjpotter** [comment]
In looking at the code closer, it seems that it can really already handle case 1.  If SkipExisting == true, then the OvfPath variable doesn't matter.  Either it will fail the template doesn't exist and when it tries to upload the ovf specified it will find its not valid or it will exist and the OvfPath will be ignored.

**Comment by avnish30jn** [comment]
Yes that is correct, if skipExisting is true then it uses the existing template. Only that the template name is generated from vm.Template and datastore name. So if the template vm is created via some other way (not libretto), then it should have the name of the datastore in the end to make it work here in libretto.

`var createTemplateName = func(t string, ds string) string {
	return fmt.Sprintf("%s-%s", t, ds)
}
.
.
.
template := createTemplateName(vm.Template, d)`
d is datastore.
